common/http/platform/beast/http_resumer.cpp

```diff
@@ -0,0 +1,454 @@
+		client_.logger_.Error(err.String());
+		client_.Cancel();
+		client_.state_.err = err;
+	} else if (content_range.range_start < client_.state_.offset) {
+		// The server resumed from an earlier offset, but we can still use the data
+		auto buffer = client_.state_.buffer.get();
+		auto overriden = client_.state_.offset - content_range.range_start;
+		client_.logger_.Warning(
+			"HTTP server returned a range with an earlier start. Expected "
+			+ to_string(client_.state_.offset) + ", got " + to_string(content_range.range_start)
+			+ ". The last " + to_string(overriden) + " bytes will be overriden");
+		buffer->erase(buffer->end() - overriden, buffer->end());
+		client_.state_.offset = buffer->size();
+	}
+}
+
+events::EventHandler BodyHandlerFunctor::GetWaitCallback() {
+	HeaderHandlerFunctor resumer_next_header_handler {
+		user_header_handler_, user_body_handler_, user_request_, client_};
+	BodyHandlerFunctor resumer_next_body_handler {
+		user_header_handler_, user_body_handler_, user_request_, client_};
+
+	return [resumer_next_header_handler,
+			resumer_next_body_handler,
+			range_req = RemainingRangeRequest()](error::Error err) {
+		if (err != error::NoError) {
+			auto err_user =
+				MakeError(DownloadResumerError, "Unexpected error in wait timer: " + err.String());
+			resumer_next_body_handler.client_.logger_.Error(err_user.String());
+			resumer_next_body_handler.client_.Cancel();
+			resumer_next_body_handler.user_body_handler_(expected::unexpected(err_user));
+			return;
+		}
+
+		auto next_call_err = resumer_next_body_handler.client_.client_.AsyncCall(
+			range_req, resumer_next_header_handler, resumer_next_body_handler);
+		if (next_call_err != error::NoError) {
+			auto err_user = MakeError(
+				DownloadResumerError,
+				"Failed to copy data to schedule the next resumer call: " + next_call_err.String());
+			resumer_next_body_handler.client_.logger_.Error(err_user.String());
+			resumer_next_body_handler.client_.Cancel();
+			resumer_next_body_handler.user_body_handler_(expected::unexpected(err_user));
+		}
+	};
+}
+
+void BodyHandlerFunctor::operator()(ExpectedIncomingResponsePtr exp_resp) {
+	// It was intentionally cancelled or otherwise the header handler errored.
+	if (!client_.IsOngoing()) {
+		if (client_.state_.err != error::NoError) {
+			user_body_handler_(expected::unexpected(client_.state_.err));
+		}
+		return;
+	}
+
+	// We can resume the download if either:
+	// * there is a short read (partial_message error) or
+	// * successful read with status code Partial Content and there is still data missing
+	const bool is_short_read =
+		!exp_resp && exp_resp.error().code == http::make_error_code(http::error::partial_message);
+	const bool is_range_response =
+		exp_resp && exp_resp.value().get()->GetStatusCode() == mender::http::StatusPartialContent;
+	const auto buffer_size = (long long int) client_.state_.buffer.get()->size();
```

Comment:
Is `size_t` better?

Reply:
I went back and forth with this and I don't know what is actually best. To use `size_t` here (and not cast again back later to `long long` for the [comparison](https://gitlab.com/Northern.tech/Mender/mender/-/jobs/4976660957)) I should then also set the fields of `struct RangeHeader` also to `size_t`.

I chose there signed version thinking that maybe a misbehaving server could send negative numbers. And also it was convenient and coherent to use `common::StringToLongLong` with we already used to parse `Content-Lenght` header.

Reply:
This discussion is obsolete now.